### PR TITLE
Add generic control blocks for model dispatch

### LINKED ISSUE #154: Implement functions for default modeling blocks
Many of the industrial models rely on common block structures we should implement as functions to make it easier to have code re-use.

### PR DESCRIPTION
**Pull request by rodrigomha**
Closes #154 

To do:
- [x] AVRs
- [x] PSSs
- [x] TGs
- [x] Outer Controls
- [x] Inner Controls
- [x] PLLs

### CODE REVIEW

File: src/models/common_controls.jl [hunk 1]
```diff
@@ -0,0 +1,268 @@
+## Common implementation of each block:
+## returns a tuple of (output, internal_derivative)
+
+"""
+Low Pass Filter
```
Comment (jd-lara) [inline src/models/common_controls.jl:5]: Limited Low pass filter

File: src/models/common_controls.jl [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+# Low pass modified with denominator K_den instead of 1.0
+"""
+Low Pass Filter Modified
+     ┌─────────────┐
+     │      K      │
+u -> │ ────────────│ -> y
+     │ K_den + sT  │
+     └─────────────┘
+"""
+function low_pass_modified(
```
Comment (jd-lara) [inline src/models/common_controls.jl:30]: This function should call the one above for consistency

**Review by jd-lara — COMMENTED** [review]
In a first pass review, we need to be careful not to repeat too much code. We could consolidate a lot of repetitive code here.

File: src/models/common_controls.jl [hunk 3]
```diff
@@ -12,21 +12,21 @@ u -> │ ────────────│ -> y
 """
 function low_pass_modified_mass_matrix(
     u::Z,
-    y::Z,
+    y::V,
     K::Float64,
     K_den::W,
     T::Float64,
-) where {W, Z <: ACCEPTED_REAL_TYPES}
+) where {V, W, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:19]: Note that this syntax is allowing `V` and `W` to be `Any`

File: src/models/common_controls.jl [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    return y, K * u - K_den * y
+end
+
+function low_pass_modified(
+    u::Z,
+    y::V,
+    K::Float64,
+    K_den::W,
+    T::Float64,
+) where {V, W, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:29]: same here

File: src/models/common_controls.jl [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+     └────────┘
+"""
+
+# Use this one if T = 0 is allowed, and let the mass matrix take care of it.
+function low_pass_mass_matrix(
+    u::Z,
+    y::V,
+    K::Float64,
+    T::Float64,
+) where {V, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:48]: and here

File: src/models/common_controls.jl [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    return low_pass_modified_mass_matrix(u, y, K, 1.0, T)
+end
+
+function low_pass(u::Z, y::V, K::Float64, T::Float64) where {V, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:52]: and here

File: src/models/common_controls.jl [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+# Does not accept T = 0
+function low_pass_nonwindup(
+    u::Z,
+    y::V,
+    K::Float64,
+    T::Float64,
+    y_min::Float64,
+    y_max::Float64,
+) where {V, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:92]: and here

File: src/models/common_controls.jl [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+function low_pass_nonwindup_ramp_limits(
+    u::Z,
+    y::V,
+    K::Float64,
+    T::Float64,
+    y_min::Float64,
+    y_max::Float64,
+    dy_min::Float64,
+    dy_max::Float64,
+) where {V, Z <: ACCEPTED_REAL_TYPES}
```
Comment (jd-lara) [inline src/models/common_controls.jl:107]: and here

File: src/models/common_controls.jl [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+function lead_lag_2nd_mass_matrix(
+    u::Z,
+    x1::Z,
+    x2::Z,
+    T1::Float64,
+    T2::Float64,
+    T3::Float64,
+    T4::Float64,
+) where {Z <: ACCEPTED_REAL_TYPES}
+    dx1dt = u - T1 * x1 - x2
```
Comment (jd-lara) [inline src/models/common_controls.jl:233]: some of these allocations aren't neccessary

File: src/models/generator_models/avr_models.jl [hunk 10]
```diff
@@ -100,18 +100,17 @@ function mdl_avr_ode!(
...
+    y_hp, dVr2_dt = high_pass(Vf, Vr2, Kf, Tf)
+    _, dVr1_dt = low_pass(V_ref - Vm - y_hp, Vr1, Ka, Ta)
+    _, dVf_dt = low_pass_modified(Vr1, Vf, 1.0, Ke + Se_Vf, Te)
 
     #Compute 4 States AVR ODE:
-    output_ode[local_ix[1]] = -(1.0 / Te) * (Vf * (Ke + Se_Vf) - Vr1) #16.12c
-    output_ode[local_ix[2]] = (1.0 / Ta) * (Ka * (V_ref - Vm - Vr2 - (Kf / Tf) * Vf) - Vr1) #16.12a
-    output_ode[local_ix[3]] = -(1.0 / Tf) * ((Kf / Tf) * Vf + Vr2) #16.12b
-    output_ode[local_ix[4]] = (1.0 / Tr) * (V_th - Vm) #16.11
+    output_ode[local_ix[1]] = dVf_dt #16.12c
```
Comment (jd-lara) [inline src/models/generator_models/avr_models.jl:110]: We should do this instead where needed. 
```suggestion
    output_ode[local_ix[1]] = low_pass_modified(Vr1, Vf, 1.0, Ke + Se_Vf, Te)[2] #16.12c
```

File: src/models/inverter_models/outer_control_models.jl [hunk 11]
```diff
@@ -132,8 +140,11 @@ function _mdl_ode_RE_active_controller_AB!(
     #Define internal states for outer control
     p_ord = active_controller_states[1]
 
+    #Compute block derivatives
+    _, dpord_dt = low_pass(p_ref, p_ord, 1.0, T_pord)
```
Comment (rodrigomha) [inline src/models/inverter_models/outer_control_models.jl:144]: This case is the only on which `p_ref` and `p_ord` can be different type that forces me to make `low_pass` inputs to accept different types. All other models the inputs and states match the type when using `low_pass` blocks

File: src/models/common_controls.jl (hunk 3 above)
Comment (rodrigomha) [reply src/models/common_controls.jl:19]: Let's resolve this later. I would prefer that Z = V, and W can potentially be different

File: src/models/inverter_models/outer_control_models.jl (hunk 11 above)
Comment (jd-lara) [reply src/models/inverter_models/outer_control_models.jl:144]: so in this case `p_ord = p_ref`

**Review by jd-lara — APPROVED** [review]
We can merge these changes and then do a more detail performance review.

**Pull request merged**